### PR TITLE
executor: check for null values when comparing different groups during streamAgg

### LINKED ISSUE #15690: executor: distinct error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
```
diff --git a/sessionctx/variable/tidb_vars.go b/sessionctx/variable/tidb_vars.go

-       DefInitChunkSize                   = 32
-       DefMaxChunkSize                    = 1024
+       DefInitChunkSize                   = 2
+       DefMaxChunkSize                    = 2

diff --git a/sessionctx/variable/varsutil.go b/sessionctx/variable/varsutil.go
-       maxChunkSizeLowerBound = 32
+       maxChunkSizeLowerBound = 1

```

download the file with all DDLs 
[randgen_index_merge_join.txt](https://github.com/pingcap/tidb/files/4381467/randgen_index_merge_join.txt)
```
source randgen_index_merge_join.txt
set @@tidb_max_chunk_size=2;

```
```
SELECT DISTINCT  table2 . `col_int_key` AS field1 FROM  P AS table1  RIGHT  JOIN P AS table2 ON  table1 . `pk` =  table2 . `pk`  WHERE table2 . `pk` <= table2 . `col_int`   ORDER BY field1  LIMIT 100 ;
```

### 2. What did you expect to see?
```
+--------+
| field1 |
+--------+
|   NULL |
|      0 |
|      2 |
|      4 |
|      8 |
+--------+
```

### 3. What did you see instead?
```
mysql> SELECT DISTINCT  table2 . `col_int_key` AS field1 FROM  P AS table1  RIGHT  JOIN P AS table2 ON  table1 . `pk` =  table2 . `pk`  WHERE table2 . `pk` <= table2 . `col_int`   ORDER BY field1  LIMIT 100 ;
+--------+
| field1 |
+--------+
|   NULL |
|      2 |
|      4 |
|      8 |
+--------+
4 rows in set (0.00 sec)

mysql> explain SELECT DISTINCT  table2 . `col_int_key` AS field1 FROM  P AS table1  RIGHT  JOIN P AS table2 ON  table1 . `pk` =  table2 . `pk`  WHERE table2 . `pk` <= table2 . `col_int`   ORDER BY field1  LIMIT 100 ;
+-----------------------------------+---------+-----------+----------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------+
| id                                | estRows | task      | access object                                | operator info                                                                                                                                   |
+-----------------------------------+---------+-----------+----------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------+
| Limit_15                          | 5.76    | root      |                                              | offset:0, count:100                                                                                                                             |
| └─StreamAgg_33                    | 5.76    | root      |                                              | group by:randgen_index_merge_join.p.col_int_key, funcs:firstrow(randgen_index_merge_join.p.col_int_key)->randgen_index_merge_join.p.col_int_key |
|   └─IndexLookUp_37                | 7.20    | root      |                                              |                                                                                                                                                 |
|     ├─IndexFullScan_34(Build)     | 9.00    | cop[tikv] | table:table2, index:col_int_key(col_int_key) | keep order:true, stats:pseudo                                                                                                                   |
|     └─Selection_36(Probe)         | 7.20    | cop[tikv] |                                              | le(randgen_index_merge_join.p.pk, randgen_index_merge_join.p.col_int)                                                                           |
|       └─TableRowIDScan_35         | 9.00    | cop[tikv] | table:table2                                 | keep order:false, stats:pseudo                                                                                                                  |
+-----------------------------------+---------+-----------+----------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------+
6 rows in set (0.01 sec)

mysql> 

```
### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
5.7.25-TiDB-v4.0.0-beta-523-g7eba696bb

### LINKED ISSUE #15691: executor, planner: IndexJoin error with limit
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
download the file with all DDLs 
[randgen_index_merge_join.txt](https://github.com/pingcap/tidb/files/4381467/randgen_index_merge_join.txt)
```
source randgen_index_merge_join.txt
```
```
SELECT   /*+INL_JOIN(table1,table2,table3,table4,table5,table6,table7,table8,table9,table10,table11,table12)*/ distinct  table2 . `col_int_key` AS field1 FROM   W AS table1  RIGHT  JOIN I AS table2 ON  table1 . `col_int_key` =  table2 . `col_int`   LEFT  JOIN N AS table3 ON  table2 . `col_varchar_1024_utf8` =  table3 . `col_varchar_1024_latin1`  WHERE table1 . `col_int` IS NOT NULL   ORDER BY field1  LIMIT 4 ;
```

### 2. What did you expect to see?
MySQL
+--------+
| field1 |
+--------+
|   NULL |
|      0 |
|      1 |
|      2 |
+--------+
### 3. What did you see instead?
+--------+
| field1 |
+--------+
|   NULL |
|   NULL |
|      0 |
|      1 |
+--------+
### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
5.7.25-TiDB-v4.0.0-beta-523-g7eba696bb

### LINKED ISSUE #15695: executor: indexJoin and indexHashJoin error with distinct
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
set the chunk size to 2 like https://github.com/pingcap/tidb/issues/15696

download the file with all DDLs 
[randgen_index_merge_join.txt](https://github.com/pingcap/tidb/files/4381467/randgen_index_merge_join.txt)
```
source randgen_index_merge_join.txt
```
```
SELECT /*+INL_JOIN(table1,table2,table3,table4,table5,table6,table7,table8,table9,table10,table11,table12)*/ distinct table1 . `col_int_key` AS field1 FROM   BB AS table1  LEFT  JOIN M AS table2 ON  table1 . `pk` =  table2 . `col_int_key`   LEFT  JOIN  M AS table3  LEFT  JOIN Y AS table4 ON  table3 . `col_varchar_10_utf8_key` =  table4 . `col_varchar_10_latin1`  ON  table2 . `col_int_key` =  table4 . `col_int_key`  WHERE table2 . `col_int_key` IS  NULL  GROUP BY field1  ORDER BY field1   ;
```

### 2. What did you expect to see?
Mysql
```
+--------+
| field1 |
+--------+
|   NULL |
|      0 |
|      1 |
|      2 |
|      3 |
|      4 |
|      5 |
|      6 |
|      7 |
|      8 |
|      9 |
+--------+
11 rows in set (0.00 sec)
```
### 3. What did you see instead?
```
mysql> SELECT /*+INL_hash_JOIN(table1,table2,table3,table4,table5,table6,table7,table8,table9,table10,table11,table12)*/ distinct table1 . `col_int_key` AS field1 FROM   BB AS table1  LEFT  JOIN M AS table2 ON  table1 . `pk` =  table2 . `col_int_key`   LEFT  JOIN  M AS table3  LEFT  JOIN Y AS table4 ON  table3 . `col_varchar_10_utf8_key` =  table4 . `col_varchar_10_latin1`  ON  table2 . `col_int_key` =  table4 . `col_int_key`  WHERE table2 . `col_int_key` IS  NULL  GROUP BY field1  ORDER BY field1   ;
+--------+
| field1 |
+--------+
|   NULL |
|      1 |
|      2 |
|      3 |
|      4 |
|      5 |
|      6 |
|      7 |
|      8 |
|      9 |
+--------+
10 rows in set, 1 warning (0.01 sec)

mysql> SELECT /*+INL_JOIN(table1,table2,table3,table4,table5,table6,table7,table8,table9,table10,table11,table12)*/ distinct table1 . `col_int_key` AS field1 FROM   BB AS table1  LEFT  JOIN M AS table2 ON  table1 . `pk` =  table2 . `col_int_key`   LEFT  JOIN  M AS table3  LEFT  JOIN Y AS table4 ON  table3 . `col_varchar_10_utf8_key` =  table4 . `col_varchar_10_latin1`  ON  table2 . `col_int_key` =  table4 . `col_int_key`  WHERE table2 . `col_int_key` IS  NULL  GROUP BY field1  ORDER BY field1   ;
+--------+
| field1 |
+--------+
|   NULL |
|      1 |
|      2 |
|      3 |
|      4 |
|      5 |
|      6 |
|      7 |
|      8 |
|      9 |
+--------+
10 rows in set, 1 warning (0.01 sec)


```
### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
5.7.25-TiDB-v4.0.0-beta-523-g7eba696bb

### LINKED ISSUE #15696: executor: indexJoin and indexHashJoin error with limit
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
Modify the source code to set 
```
diff --git a/sessionctx/variable/tidb_vars.go b/sessionctx/variable/tidb_vars.go

-       DefInitChunkSize                   = 32
-       DefMaxChunkSize                    = 1024
+       DefInitChunkSize                   = 2
+       DefMaxChunkSize                    = 2

diff --git a/sessionctx/variable/varsutil.go b/sessionctx/variable/varsutil.go
-       maxChunkSizeLowerBound = 32
+       maxChunkSizeLowerBound = 1

```

related #15691 
download the file with all DDLs 
[randgen_index_merge_join.txt](https://github.com/pingcap/tidb/files/4381467/randgen_index_merge_join.txt)

```
source randgen_index_merge_join.txt
 set @@tidb_max_chunk_size=2;
```
```
SELECT  /*+INL_hash_JOIN(table1,table2,table3,table4,table5,table6,table7,table8,table9,table10,table11,table12)*/ table1 . `col_int_key` AS field1 FROM  CC AS table1  LEFT  JOIN Z AS table2 ON  table1 . `col_varchar_1024_utf8_key` =  table2 . `col_varchar_1024_latin1_key`  WHERE table2 . `pk` >= 6  GROUP BY field1  ORDER BY field1  LIMIT 10;

```

### 2. What did you expect to see?
mysql
```
+--------+
| field1 |
+--------+
|   NULL |
|      0 |
|      1 |
|      2 |
|      3 |
|      4 |
|      5 |
|      6 |
|      7 |
|      8 |
+--------+
10 rows in set (0.03 sec)
```
### 3. What did you see instead?
```
mysql> SELECT  /*+INL_JOIN(table1,table2,table3,table4,table5,table6,table7,table8,table9,table10,table11,table12)*/ table1 . `col_int_key` AS field1 FROM  CC AS table1  LEFT  JOIN Z AS table2 ON  table1 . `col_varchar_1024_utf8_key` =  table2 . `col_varchar_1024_latin1_key`  WHERE table2 . `pk` >= 6  GROUP BY field1  ORDER BY field1  LIMIT 10; 
+--------+
| field1 |
+--------+
|   NULL |
|   NULL |
|   NULL |
|      0 |
|      1 |
|      2 |
|      3 |
|      4 |
|      5 |
|      6 |
+--------+
10 rows in set, 1 warning (0.02 sec)

mysql> SELECT  /*+INL_hash_JOIN(table1,table2,table3,table4,table5,table6,table7,table8,table9,table10,table11,table12)*/ table1 . `col_int_key` AS field1 FROM  CC AS table1  LEFT  JOIN Z AS table2 ON  table1 . `col_varchar_1024_utf8_key` =  table2 . `col_varchar_1024_latin1_key`  WHERE table2 . `pk` >= 6  GROUP BY field1  ORDER BY field1  LIMIT 10;
+--------+
| field1 |
+--------+
|   NULL |
|   NULL |
|   NULL |
|      0 |
|      1 |
|      2 |
|      3 |
|      4 |
|      5 |
|      6 |
+--------+
10 rows in set, 1 warning (0.02 sec)

```

similar problem for 
```
SELECT DISTINCT  /*+INL_MERGE_JOIN(table1,table2,table3,table4,table5,table6,table7,table8,table9,table10,table11,table12)*/ table2 . `col_int_key` AS field1 FROM  Q AS table1  RIGHT  JOIN   B AS table2  LEFT  JOIN S AS table3 ON  table2 . `col_int` =  table3 . `col_int_key`   LEFT  JOIN A AS table4 ON  table2 . `pk` =  table4 . `pk`  ON  table1 . `pk` =  table3 . `pk`  WHERE table3 . `col_int` IS  NULL  GROUP BY field1  ORDER BY field1  LIMIT 10 ;
```
```
SELECT DISTINCT  /*+INL_MERGE_JOIN(table1,table2,table3,table4,table5,table6,table7,table8,table9,table10,table11,table12)*/ table1 . `col_int_key` AS field1 FROM   R AS table1  LEFT  JOIN W AS table2 ON  table1 . `col_varchar_10_latin1` =  table2 . `col_varchar_10_latin1_key`   LEFT OUTER JOIN P AS table3 ON  table2 . `col_varchar_1024_utf8_key` =  table3 . `col_varchar_1024_latin1`  WHERE table2 . `col_int` < 6  GROUP BY field1  ORDER BY field1  LIMIT 2 ;
```
### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
5.7.25-TiDB-v4.0.0-beta-523-g7eba696bb

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?
Close #15690 
Close #15691
Close #15695 
Close #15696
In the master version, we forget to check the null value when we compare the different groups during `StreamAgg`.

### What is changed and how it works?
Add checks for null values when comparing different groups during `StreamAgg`.

### Check List 

Tests 

- Unit test
- Integration test

### CODE REVIEW

**Comment by qw4990** [comment]
Do we need to cherry-pick this PR to release-4.0?

File: executor/aggregate_test.go [hunk 1]
```diff
@@ -844,3 +844,13 @@ func (s *testSuiteAgg) TestPR15242ShallowCopy(c *C) {
...
 }
+
+func (s *testSuiteAgg) TestIssue15690(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec(`drop table if exists t;`)
+	tk.MustExec(`create table t(a int);`)
+	tk.MustExec(`insert into t values(null),(null);`)
+	tk.MustExec(`insert into t values(0),(2),(2),(4),(8);`)
+	tk.Se.GetSessionVars().MaxChunkSize = 2
+	tk.MustQuery(`select /*+ stream_agg() */ distinct * from t;`).Check(testkit.Rows("<nil>", "0", "2", "4", "8"))
```
Comment (qw4990) [inline executor/aggregate_test.go]: Almost LGTM, but please add more tests to cover more types for safety.

**Comment by Reminiscent** [comment]
@SunRunAway @qw4990 @XuHuaiyu  Update. PTAL

**Comment by Reminiscent** [comment]
> Could you add some test cases for `splitIntoGroups` like others https://github.com/search?q=splitIntoGroups+repo%3Apingcap%2Ftidb+filename%3A_test.go&type=Code&ref=advsearch&l=&l=?

https://github.com/pingcap/tidb/blob/7746fbd880bb19f40492dda286f9eb82c87d78c7/executor/executor_required_rows_test.go#L814-L816  Here can check the result of  `splitIntoGroups` .

File: executor/aggregate_test.go [hunk 2]
```diff
@@ -844,3 +844,48 @@ func (s *testSuiteAgg) TestPR15242ShallowCopy(c *C) {
...
+
+func (s *testSuiteAgg) TestIssue15690(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.Se.GetSessionVars().MaxChunkSize = 2
+	// check for INT type
+	tk.MustExec(`drop table if exists t;`)
+	tk.MustExec(`create table t(a int);`)
+	tk.MustExec(`insert into t values(null),(null);`)
+	tk.MustExec(`insert into t values(0),(2),(2),(4),(8);`)
+	tk.MustQuery(`select /*+ stream_agg() */ distinct * from t;`).Check(testkit.Rows("<nil>", "0", "2", "4", "8"))
```
Comment (SunRunAway) [inline executor/aggregate_test.go]: better to ensure there's no output by executing `show warning` if hint is used here.

File: executor/aggregate.go [hunk 3]
```diff
@@ -1111,8 +1111,15 @@ func (e *vecGroupChecker) evalGroupItemsAndResolveGroups(item expression.Express
...
 
-	previousIsNull := col.IsNull(0)
 	var firstRowDatum, lastRowDatum types.Datum
+	firstRowIsNull, lastRowIsNull := col.IsNull(0), col.IsNull(numRows-1)
+	if firstRowIsNull {
+		firstRowDatum.SetNull()
+	}
+	if lastRowIsNull {
+		lastRowDatum.SetNull()
+	}
```
Comment (fzhedu) [inline executor/aggregate.go]: move this behand switch eType, and reduce the if to check null in the switch.

**Pull request merged**